### ISSUE TITLE
What would a generic train operation look like?

### ISSUE THREAD

**Issue by umesh-timalsina** [deepforge-dev/deepforge#1689]
It might be worth investigating a way to have a generic train operation. Not sure whether it should be there. But a generic train operation can be easily added for `keras` library. If to be added where should be added and what route should we take?

**Comment by brollb** [comment]
It sounds like there are a few questions here so I am going to try to answer each directly (let me know if I miss anything!):

**Should a generic training operation be added?**
I think it would be *really* valuable to add. A generic training operation would make it much easier for people to get started quickly. 

**Where should it be added?**
Good question. Ideally, we would have some publicly available libraries that are able to be imported from within DeepForge (similar to importing extensions). WebGME has a mechanism for importing libraries but I do not know of any mechanism for adding these libraries to a registry to make them discoverable. This would be a good discussion to start with the WebGME community as it may be of interest to WebGME itself. If not, we can add this registry as a router in DeepForge. There are other considerations that would need to be discussed like private vs public libraries (ie, sharing between a single user's projects).

**What route should we take?**
There are two separate fronts to work on: creating the operations to include in the library and the library infrastructure to aid in discovery and distribution.
The first front is going to be the focus of the summer intern (not just with a training operation but it would be good to include). I would recommend starting with the existing `TrainValidate` operation from the redshift example then removing the code specific to the given example such as:
```python
        X = dataset['X']
        y = dataset['y']
        y_cats = self.to_categorical(y)  # This is the most concerning
        # ... generic training code

    def to_categorical(self, y, max_y=0.4, num_possible_classes=32):
        one_step = max_y / num_possible_classes
        y_cats = []
        for values in y:
            y_cats.append(int(values[0] / one_step))
        return y_cats
```
and parameterizing the operation by more of the generic parameters such as `optimizer`, `loss` and `metrics`. These would probably be best suited as an enum on the operation. This will also require some modifications of things like the generated plots (which should certainly exist). Finally, the rest of the examples should be updated to use this standard `Train` operation. This would also serve as a nice demonstration of the general nature of the operation.

On the second front, we should first figure out if it should be added in DeepForge or contributed back to WebGME. It seems pretty generic so I suspect it would be a good candidate for contributing back. We will need to make sure we address a few questions/considerations:
- public vs private/personal libraries
- library dependencies. Libraries may require specific DeepForge extensions to be available (such as DeepForge-Keras) whereas others may be generic and may not require any extensions.
- categories/labels for libraries. This may not be necessary initially but it would be useful in the future to be able to sort by things like "image processing", "astronomy", etc.

**Comment by brollb** [comment]
This would actually be better to implement as an interactive editor (#1808) as this will let the user get feedback immediately, use a more sophisticated UI (especially for customizing optimizers, etc) and will still generate a pipeline at the end of the day.

Closing this in favor of #1816.
